File: server/ops/diagnostics/source.js
// The Heartbeat source that actually runs registered diagnostic checks —
// same registerSource({id, defaultIntervalMs, listItems, check}) contract
// every other source uses. This is where retry-then-escalate (the SAME
// shape Jobs already built — one automatic attempt, then tell the owner
// plainly) gets applied to self-diagnosis specifically.
//
// **One remedy attempt per NEW failure occurrence, never per tick** — the
// same "canAutoRetry: retries===0" discipline job-policy.js already
// established, adapted here via checkState.lastOutcome instead of a
// numeric retry counter: a check that's ALREADY marked 'failing' from the
// previous tick has already had its one remedy shot (or had none to try),
// so a still-failing check on tick N+1 does not retry again and does not
// re-report — only a genuine ok->failing TRANSITION triggers the
// retry-then-escalate sequence. Own-checkState dedup, same discipline
// heartbeat/CLAUDE.md documents at length as a real, previously-live bug
// (jobs-source.js) — required here for the identical reason: a sustained
// failure must not re-spend a decision.js model call and re-notify on
// every single tick forever.
//
// Every probe/remedy/outcome writes a real ops_trace row
// (source:'diagnosis', sourceRef: check.id) — this is what makes "has
// anything gone wrong with you lately?" a real, queryable answer rather
// than a claim with nothing behind it.

import { listChecks, getCheck } from './registry.js';
import { getItem as getScheduleItem } from '../../heartbeat/schedule-store.js';
import { appendEntry as appendTrace } from '../ops-trace.js';

export const SOURCE_ID = 'diagnosis';
const CHECK_INTERVAL_MS = 5 * 60 * 1000; // cheap probes only — real network/DB reads, no model call of their own

export async function listItems() {
  return listChecks().map((c) => ({ itemKey: c.id, intervalMs: CHECK_INTERVAL_MS }));
}

function traceIntent(checkId, kind, summary, detail) {
  try {
    appendTrace({ source: 'diagnosis', sourceRef: checkId, phase: 'intent', effect: 'read', kind, summary, detail });
  } catch (err) {
    console.error(`[ops] diagnosis trace (intent) failed for "${checkId}":`, err);
  }
}
function traceOutcome(checkId, kind, summary, detail) {
  try {
    appendTrace({ source: 'diagnosis', sourceRef: checkId, phase: 'outcome', effect: 'read', kind, summary, detail });
  } catch (err) {
    console.error(`[ops] diagnosis trace (outcome) failed for "${checkId}":`, err);
  }
}

export async function check(itemKey) {
  const c = getCheck(itemKey);
  if (!c) return { finding: null, checkState: null }; // check was un-registered since this item was scheduled — nothing to report

  const prior = getScheduleItem(SOURCE_ID, itemKey)?.checkState;
  const wasFailing = prior?.lastOutcome === 'failing';

  traceIntent(itemKey, 'probe', `Running diagnostic check: ${c.id}`);
  let result;
  try {
    result = await c.probe();
  } catch (err) {
    result = { ok: false, detail: `Probe threw: ${err?.message || err}` };
  }
  traceOutcome(itemKey, 'probe', result.ok ? 'Probe passed.' : `Probe failed: ${result.detail || 'no detail given'}`, result.detail || null);

  if (result.ok) {
    if (wasFailing) {
      // Recovered — either self-healed by a remedy on a prior tick, or
      // fixed some other way. Worth a trace row either way; not worth
      // interrupting the owner about a problem that's already gone.
      traceOutcome(itemKey, 'note', `"${c.id}" recovered.`);
    }
    return { finding: null, checkState: { lastOutcome: 'ok' } };
  }

  // Failing. Only a genuine NEW occurrence gets a remedy attempt — a check
  // still failing from a previous tick already had its one shot.
  if (wasFailing) {
    return { finding: null }; // already attempted-and/or-reported this exact ongoing failure — checkState left untouched
  }

  if (typeof c.remedy === 'function') {
    traceIntent(itemKey, 'remedy', `Attempting the one automatic remedy for "${c.id}".`);
    try {
      await c.remedy();
    } catch (err) {
      traceOutcome(itemKey, 'remedy', `Remedy threw: ${err?.message || err}`);
    }
    let reprobe;
    try {
      reprobe = await c.probe();
    } catch (err) {
      reprobe = { ok: false, detail: `Re-probe threw: ${err?.message || err}` };
    }
    if (reprobe.ok) {
      traceOutcome(itemKey, 'remedy', `"${c.id}" self-healed after one remedy attempt.`);
      return { finding: null, checkState: { lastOutcome: 'ok' } };
    }
    traceOutcome(itemKey, 'decision', `"${c.id}" still failing after the one automatic remedy attempt — escalating.`, reprobe.detail || null);
    return {
      finding: { summary: `Self-diagnosis: "${c.id}" is failing, and the one automatic remedy attempt didn't fix it.`, detail: reprobe.detail || result.detail || null },
      checkState: { lastOutcome: 'failing' },
    };
  }

  // No remedy declared — escalate straight from the failed probe.
  traceOutcome(itemKey, 'decision', `"${c.id}" is failing with no remedy available — escalating.`, result.detail || null);
  return {
    finding: { summary: `Self-diagnosis: "${c.id}" is failing.`, detail: result.detail || null },
    checkState: { lastOutcome: 'failing' },
  };
}

export const source = { id: SOURCE_ID, defaultIntervalMs: CHECK_INTERVAL_MS, listItems, check };

File: backend/jarvis/ops/diagnostics/source.py
"""Running the registered checks, and the one automatic attempt at a fix.

**One remedy per NEW failure, never per tick.** Only an ok-to-failing transition
spends a remedy and produces a finding; a check still failing on the next tick
does neither. Without that, one broken thing means a remedy attempt, a model
call to judge urgency and a notification every few minutes for as long as it
stays broken — which is both expensive and the fastest way to train someone to
ignore notifications.

The state that makes this possible is the schedule row's own `checkState`, the
same mechanism every other source uses for the same reason.

Every probe, remedy and outcome writes a trace row, which is what makes "has
anything gone wrong with you lately?" a real answer rather than a claim.
"""

from __future__ import annotations

import logging
from typing import Any

from ...heartbeat.sources.registry import Source
from .. import trace as ops_trace
from .registry import get_check, list_checks

logger = logging.getLogger(__name__)

SOURCE_ID = "diagnosis"
#: Cheap probes only — real reads, no model call of their own.
CHECK_INTERVAL_MS = 5 * 60 * 1000


def _trace(check_id: str, phase: str, kind: str, summary: str, detail: Any = None) -> None:
    try:
        ops_trace.append(source=SOURCE_ID, source_ref=check_id, phase=phase, effect="read",
                         kind=kind, summary=summary, detail=detail)
    except Exception:  # noqa: BLE001 — a trace failure must not fail the diagnosis
        logger.exception("could not trace diagnosis of %s", check_id)


def list_items() -> list[dict[str, Any]]:
    return [{"itemKey": check.id, "intervalMs": CHECK_INTERVAL_MS} for check in list_checks()]


def _probe(check: Any) -> dict[str, Any]:
    try:
        return check.probe() or {"ok": False, "detail": "The check returned nothing."}
    except Exception as err:  # noqa: BLE001 — a check that throws IS a failing check
        return {"ok": False, "detail": f"The check itself failed: {err}"}


def check(item_key: str) -> dict[str, Any]:
    from ...heartbeat import schedule_store

    registered = get_check(item_key)
    if registered is None:
        # Unregistered since this item was scheduled: nothing to run, and the
        # row is pruned on the next reconcile.
        return {"finding": None, "checkState": None}

    item = schedule_store.get_item(SOURCE_ID, item_key)
    was_failing = ((item or {}).get("checkState") or {}).get("lastOutcome") == "failing"

    _trace(item_key, "intent", "probe", f"Checking: {item_key}")
    result = _probe(registered)
    _trace(item_key, "outcome", "probe",
           "Passed." if result.get("ok") else f"Failed: {result.get('detail') or 'no detail'}",
           result.get("detail"))

    if result.get("ok"):
        if was_failing:
            _trace(item_key, "outcome", "note", f'"{item_key}" is working again.')
        return {"finding": None, "checkState": {"lastOutcome": "ok"}}

    if was_failing:
        # Already had its one attempt and was already reported. Nothing new has
        # happened, so nothing new is said.
        return {"finding": None}

    if registered.remedy is not None:
        _trace(item_key, "intent", "remedy", f'Trying the one automatic fix for "{item_key}".')
        try:
            registered.remedy()
        except Exception as err:  # noqa: BLE001
            _trace(item_key, "outcome", "remedy", f"The fix itself failed: {err}")
        again = _probe(registered)
        if again.get("ok"):
            _trace(item_key, "outcome", "remedy", f'"{item_key}" fixed itself after one attempt.')
            return {"finding": None, "checkState": {"lastOutcome": "ok"}}
        _trace(item_key, "outcome", "decision",
               f'"{item_key}" is still failing after the one automatic fix.', again.get("detail"))
        return {
            "finding": {"summary": f'Something is wrong with me: "{item_key}" is failing, and '
                                   f"trying to fix it automatically did not work.",
                        "detail": again.get("detail") or result.get("detail")},
            "checkState": {"lastOutcome": "failing"},
        }

    _trace(item_key, "outcome", "decision",
           f'"{item_key}" is failing and has no automatic fix.', result.get("detail"))
    return {"finding": {"summary": f'Something is wrong with me: "{item_key}" is failing.',
                        "detail": result.get("detail")},
            "checkState": {"lastOutcome": "failing"}}


source = Source(id=SOURCE_ID, default_interval_ms=CHECK_INTERVAL_MS,
                list_items=list_items, check=check)


def recent_health(hours: int = 24, limit: int = 200) -> dict[str, Any]:
    """What the checks have actually found lately — the pull path's raw material."""
    from datetime import datetime, timedelta, timezone

    from ...jscompat import to_iso_z

    since = to_iso_z(datetime.now(timezone.utc) - timedelta(hours=hours))
    rows = ops_trace.recent(SOURCE_ID, since_iso=since, limit=limit)
    failing: dict[str, str] = {}
    from ...heartbeat import schedule_store
    for item in schedule_store.list_for_source(SOURCE_ID):
        if ((item.get("checkState") or {}).get("lastOutcome")) == "failing":
            failing[item["itemKey"]] = "failing"
    return {
        "hours": hours,
        "checksRegistered": [c.id for c in list_checks()],
        "currentlyFailing": sorted(failing),
        "recent": [{"check": r["source_ref"], "phase": r["phase"], "kind": r["kind"],
                    "summary": r["summary"], "at": r["created_at"]} for r in rows],
    }
